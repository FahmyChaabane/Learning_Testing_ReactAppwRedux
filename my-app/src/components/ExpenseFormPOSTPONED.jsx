import React, { useEffect, useState } from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";

const ExpenseForm = (props) => {
  const [description, setDescription] = useState("");
  const [note, setNote] = useState("");
  const [amount, setAmount] = useState("");
  const [createdAt, setCreatedAt] = useState(moment());
  const [calendarFocused, setCalendarFocused] = useState(false);
  const [error, setError] = useState("");

  useEffect(() => {
    if (props.expense) {
      console.log(props.expense);
      setDescription(props.expense.description);
      setNote(props.expense.note);
      setAmount((props.expense.amount / 100).toString());
      setCreatedAt(moment(props.expense.createdAt));
    }
  }, []);

  const onDescriptionChange = (e) => {
    const description = e.target.value;
    setDescription(description);
  };
  const onNoteChange = (e) => {
    const note = e.target.value;
    setNote(note);
  };
  const onAmountChange = (e) => {
    const amount = e.target.value;

    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      setAmount(amount);
    }
  };

  const onDateChange = (createdAt) => {
    if (createdAt) {
      setCreatedAt(createdAt);
    }
  };
  const onFocusChange = ({ focused }) => {
    setCalendarFocused(focused);
  };
  const onSubmit = (e) => {
    e.preventDefault();

    if (!description || !amount) {
      setError("Please provide description and amount.");
    } else {
      setError("");
      props.onSubmit({
        description: description,
        amount: parseFloat(amount, 10) * 100,
        createdAt: createdAt.valueOf(),
        note: note,
      });
    }
  };
  return (
    <div>
      {error && <p>{error}</p>}
      <form onSubmit={onSubmit}>
        <input
          type="text"
          placeholder="Description"
          autoFocus
          value={description}
          onChange={onDescriptionChange}
        />
        <input
          type="text"
          placeholder="Amount"
          value={amount}
          onChange={onAmountChange}
        />
        <SingleDatePicker
          date={createdAt}
          onDateChange={onDateChange}
          focused={calendarFocused}
          onFocusChange={onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
        <textarea
          placeholder="Add a note for your expense (optional)"
          value={note}
          onChange={onNoteChange}
        ></textarea>
        <button>Add Expense</button>
      </form>
    </div>
  );
};

export default ExpenseForm;
