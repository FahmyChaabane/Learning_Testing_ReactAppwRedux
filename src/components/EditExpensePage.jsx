import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { startEditExpense, startRemoveExpense } from "../actions/expenses";

export const EditExpensePage = (props) => {
  if (!props.expense) {
    return <p>id doesn't exist</p>;
  }
  return (
    <div>
      <ExpenseForm
        expense={props.expense}
        onSubmit={(expense) => {
          props.startEditExpense(props.expense.id, expense);
          props.history.push("/dashboard");
        }}
      />
      <button
        onClick={() => {
          props.startRemoveExpense({ id: props.expense.id });
          props.history.push("/dashboard");
        }}
      >
        Remove
      </button>
    </div>
  );
};

const mapDispatchToProps = (dispatch, props) => {
  return {
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpense: (data) => dispatch(startRemoveExpense(data)),
  };
};

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(
      (expense) => expense.id === props.match.params.id
    ),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
