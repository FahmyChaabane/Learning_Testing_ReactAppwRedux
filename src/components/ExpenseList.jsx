import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from "../selectors/expenses";
import total from "../selectors/expenses-total";

export const ExpenseList = (props) => (
  <div>
    <h1>Expense List</h1>
    {props.expenses.length === 0 ? (
      <span> there are no expenses </span>
    ) : (
      props.expenses.map((expense) => {
        return <ExpenseListItem key={expense.id} {...expense} />;
      })
    )}
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters),
  };
};

export default connect(mapStateToProps)(ExpenseList);
