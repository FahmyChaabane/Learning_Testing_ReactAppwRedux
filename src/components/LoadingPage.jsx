import React from "react";

const LoadingPage = () => {
  return (
    <div>
      <img src="/images/loader.gif" />
    </div>
  );
};

export default LoadingPage;
