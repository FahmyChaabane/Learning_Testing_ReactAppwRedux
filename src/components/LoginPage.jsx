import React from "react";
import { connect } from "react-redux";
import { startLogin } from "../actions/auth";

export const LoginPage = (props) => {
  return <button onClick={props.startLogin}>Login</button>;
};

const mapDispatchToProps = (dispatch, { history }) => {
  return {
    startLogin: () => dispatch(startLogin(history)),
  };
};

export default connect(undefined, mapDispatchToProps)(LoginPage);
