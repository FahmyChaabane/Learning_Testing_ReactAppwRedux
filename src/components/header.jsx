import React from "react";
import { NavLink } from "react-router-dom";
import { connect } from "react-redux";
import { startLogout } from "../actions/auth";

export const Header = (props) => (
  <header>
    <h1>Expensifie</h1>
    <NavLink to="/dashboard">Dashboard</NavLink>
    <NavLink to="/create">Create Expense</NavLink>
    <NavLink to="/help">Help</NavLink>
    <button onClick={props.startLogout}>logout</button>
  </header>
);

const mapDispatchToProps = (dispatch) => {
  return {
    startLogout: () => dispatch(startLogout()),
  };
};

export default connect(undefined, mapDispatchToProps)(Header);
