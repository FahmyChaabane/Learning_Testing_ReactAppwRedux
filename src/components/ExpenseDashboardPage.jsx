import React, { useEffect, useState } from "react";
import { connect } from "react-redux";
import ExpenseList from "./ExpenseList";
import ExpenseListFilters from "./ExpenseListFilters";
import ExpensesSummary from "./ExpensesSummary";
import { startSetExpenses } from "../actions/expenses";

export const ExpenseDashboardPage = (props) => {
  const [data, setData] = useState(true);

  useEffect(() => {
    props.setExpenses().then((expenses) => {
      if (expenses.length !== 0) setData(false);
    });
  }, []);

  return (
    <div>
      {data ? (
        <p>loading...</p>
      ) : (
        <div>
          <ExpensesSummary />
          <ExpenseListFilters />
          <ExpenseList />
        </div>
      )}
    </div>
  );
};

const mapDispatchToProps = (dispatch) => {
  return {
    setExpenses: () => dispatch(startSetExpenses()),
  };
};

export default connect(undefined, mapDispatchToProps)(ExpenseDashboardPage);
