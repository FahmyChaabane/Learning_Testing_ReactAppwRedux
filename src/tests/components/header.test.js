import React from "react";
import { shallow } from "enzyme"; // shallow renderer
//import toJson from "enzyme-to-json";
//import ReactShallowRenderer from "react-test-renderer/shallow";
import { Header } from "../../components/header";

test("should render header correctly", () => {
  const wrapper = shallow(<Header />);
  //expect(wrapper.find("h1").text()).toBe("Expensifie");
  expect(wrapper).toMatchSnapshot();
  // const renderer = new ReactShallowRenderer();
  // renderer.render(<Header />);
  // expect(renderer.getRenderOutput()).toMatchSnapshot();
  // console.log(renderer.getRenderOutput());
});

test("should call startLogout on button click", () => {
  const logOutSpy = jest.fn();
  const wrapper = shallow(<Header startLogout={logOutSpy} />);
  wrapper.find("button").simulate("click");
  expect(logOutSpy).toHaveBeenCalled();
});
