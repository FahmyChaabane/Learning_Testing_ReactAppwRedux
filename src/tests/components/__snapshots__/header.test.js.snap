// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render header correctly 1`] = `
<header>
  <h1>
    Expensifie
  </h1>
  <NavLink
    to="/dashboard"
  >
    Dashboard
  </NavLink>
  <NavLink
    to="/create"
  >
    Create Expense
  </NavLink>
  <NavLink
    to="/help"
  >
    Help
  </NavLink>
  <button>
    logout
  </button>
</header>
`;
